File: pset6/credit/credit.py
def luhn_checksum(card_number):
    # Check Luhn checksum
    def digits_of(n):
        return [int(d) for d in str(n)]
    digits = digits_of(card_number)
    odd_digits = digits[-1::-2]
    even_digits = digits[-2::-2]
    checksum = 0
    checksum += sum(odd_digits)
    for d in even_digits:
        checksum += sum(digits_of(d * 2))
    return checksum % 10


def is_luhn_valid(card_number):
    # Validate Credit Card number using Luhn’s Algorithm
    return luhn_checksum(card_number) == 0


def check_credit_card_number_length(credit_card_number):
    # Check credit card number length
    credit_card_number_length = len(str(credit_card_number))
    if credit_card_number_length in (13, 15, 16):
        return True
    return False


def get_credit_card_issuers_name(credit_card_number):
    # Get credit card issuer's name
    if check_credit_card_number_length(credit_card_number) and is_luhn_valid(credit_card_number) and credit_card_number.isdigit():
        company_id = int(str(credit_card_number)[: 2])
        if company_id in (34, 37):
            return 'AMEX'
        elif company_id in (51, 52, 53, 54, 55):
            return 'MASTERCARD'
        elif(company_id - company_id % 10) / 10 == 4:
            return 'VISA'
    return 'INVALID'


# Request credit card number input
credit_card_number = input('Number: ')

# Print result to the screen
print(get_credit_card_issuers_name(credit_card_number))

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

string getCardIssuersName(long creditCardNumber);
int luhnCheck(const char *cc);
int checkCreditCardNumberLength(long creditCardNumber);

int main(void)
{
    // Define variables
    long creditCardNumber;
    char buffer[20];

    // Ask user to input his credit card number
    creditCardNumber = get_long("Number: ");
    sprintf(buffer, "%lu", creditCardNumber);

    // Validate Credit Card number using Luhn’s Algorithm
    if (!luhnCheck(buffer) || !checkCreditCardNumberLength(creditCardNumber))
    {
        printf("INVALID\n");
    }
    // If Credit Card number is Valid print
    // Credit Card's issuer name
    else
    {
        printf("%s\n", getCardIssuersName(creditCardNumber));
    }
}

string getCardIssuersName(long creditCardNumber)
{
    long tempNum = creditCardNumber;
    int companyId;

    // Get first two digits from Credit Card number
    while (tempNum >= 100)
    {
        tempNum = tempNum / 10;
        companyId = tempNum;
    }

    // Return Credit Card's issuer name
    if (companyId == 34 || companyId == 37)
    {
        return "AMEX";
    }
    else if (companyId == 51 || companyId == 52 || companyId == 53 || companyId == 54 || companyId == 55)
    {
        return "MASTERCARD";
    }
    else if ((companyId - companyId % 10) / 10 == 4)
    {
        return "VISA";
    }
    else
    {
        return "INVALID";
    }
}

int checkCreditCardNumberLength(long creditCardNumber)
{
    if (creditCardNumber == 0)
    {
        return 0;
    }
    int nDigits = floor(log10(labs(creditCardNumber))) + 1;

    // Validate CC number length
    // American Express - 15-digit numbers, MasterCard - 16-digit numbers,
    // and Visa - 13 and 16-digit numbers
    if (nDigits != 13 && nDigits != 15 && nDigits != 16)
    {
        return 0;
    }
    return 1;
}

// Luhn's Algorithm implementation from rosettacode.org
// https://rosettacode.org/wiki/Luhn_test_of_credit_card_numbers#C
int luhnCheck(const char *cc)
{
    const int m[] = {0, 2, 4, 6, 8, 1, 3, 5, 7, 9};
    int i, odd = 1, sum = 0;

    for (i = strlen(cc); i--; odd = !odd)
    {
        int digit = cc[i] - '0';
        sum += odd ? digit : m[digit];
    }

    return sum % 10 == 0;
}